zipkin-storage/elasticsearch-http/src/main/java/zipkin/storage/elasticsearch/jest/RestClient.java

```diff
@@ -0,0 +1,397 @@
+      indices = allIndices;
+    }
+    SearchSourceBuilder elasticQuery = new SearchSourceBuilder().query(query).size(0);
+
+    for (AbstractAggregationBuilder aggregation : aggregations) {
+      elasticQuery.aggregation(aggregation);
+    }
+
+    return transform(toGuava(
+        new LenientSearch(new Search.Builder(elasticQuery.toString())
+            .addIndex(Arrays.asList(indices))
+            .addType(SPAN))
+    ), AsRestBuckets.INSTANCE);
+  }
+
+  private enum AsRestBuckets implements Function<SearchResult, Buckets> {
+    INSTANCE;
+
+    @Override public Buckets apply(SearchResult input) {
+      return new RestBuckets(input);
+    }
+  }
+
+  @Override
+  public ListenableFuture<List<Span>> findSpans(String[] indices, QueryBuilder query) {
+    if (indices.length > MAX_INDICES) {
+      query = QueryBuilders.indicesQuery(query, indices).noMatchQuery("none");
+      indices = allIndices;
+    }
+
+    return transform(toGuava(
+        new LenientSearch(new Search.Builder(new SearchSourceBuilder().query(query).size(
+            InternalElasticsearchClient.MAX_RAW_SPANS).toString())
+            .addIndex(Arrays.asList(indices))
+            .addType(SPAN))), new Function<SearchResult, List<Span>>() {
+      @Override public List<Span> apply(SearchResult input) {
+        if (input.getTotal() == 0) return null;
+
+        ImmutableList.Builder<Span> builder = ImmutableList.builder();
+        for (SearchResult.Hit<Span, ?> hit : input.getHits(Span.class)) {
+          builder.add(hit.source);
+        }
+        return builder.build();
+      }
+    });
+  }
+
+  @Override
+  public ListenableFuture<Collection<DependencyLink>> findDependencies(String[] indices) {
+    QueryBuilder query = QueryBuilders.matchAllQuery();
+    if (indices.length > MAX_INDICES) {
+      query = QueryBuilders.indicesQuery(query, indices).noMatchQuery("none");
+      indices = allIndices;
+    }
+
+    Search.Builder search = new Search.Builder(new SearchSourceBuilder().query(query).toString())
+        .addIndex(Arrays.asList(indices))
+        .addType(DEPENDENCY_LINK);
+
+    return transform(toGuava(new LenientSearch(search)),
+        new Function<SearchResult, Collection<DependencyLink>>() {
+          @Override public Collection<DependencyLink> apply(SearchResult input) {
+            ImmutableList.Builder<DependencyLink> builder = ImmutableList.builder();
+            for (SearchResult.Hit<DependencyLink, ?> hit : input.getHits(DependencyLink.class)) {
+              builder.add(hit.source);
+            }
+            return builder.build();
+          }
+        });
+  }
+
+  @Override
+  public ListenableFuture<Void> indexSpans(List<IndexableSpan> spans) {
+    if (spans.isEmpty()) return Futures.immediateFuture(null);
+
+    // Create a bulk request when there is more than one span to store
+    ListenableFuture<?> future;
+    final Set<String> indices = new HashSet<>();
```

Comment:
nit: please use LinkedHashSet/Map always :) order bugs in tests aren't worth tracking down, and such things often happen well after the first commit.